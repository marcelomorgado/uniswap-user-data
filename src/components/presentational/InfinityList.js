import React from "react";
import PropTypes from "prop-types";
import { FixedSizeList as List } from "react-window";
import InfiniteLoader from "react-window-infinite-loader";

// Component based on the react-window-infinite-loader documentation
// See more:
// https://github.com/bvaughn/react-window-infinite-loader
// https://codesandbox.io/s/x70ly749rq
function InfiniteList({
  hasNextPage,
  isNextPageLoading,
  items,
  loadNextPage,
  rowRender,
  rowHeight,
}) {
  const itemCount = hasNextPage ? items.length + 1 : items.length;

  const loadMoreItems = isNextPageLoading ? () => {} : loadNextPage;

  const isItemLoaded = index => !hasNextPage || index < items.length;

  // Render an item or a loading indicator.
  const Item = ({ index, style }) => {
    if (!isItemLoaded(index)) {
      return <div style={style}>{"Loading..."}</div>;
    }
    const item = items[index];
    return rowRender({ item, style });
  };

  Item.propTypes = {
    index: PropTypes.number.isRequired,
    style: PropTypes.object.isRequired,
  };

  return (
    <InfiniteLoader
      isItemLoaded={isItemLoaded}
      itemCount={itemCount}
      loadMoreItems={loadMoreItems}
    >
      {({ onItemsRendered, ref }) => (
        <List
          className="List"
          height={600}
          itemCount={itemCount}
          itemSize={rowHeight}
          onItemsRendered={onItemsRendered}
          ref={ref}
        >
          {Item}
        </List>
      )}
    </InfiniteLoader>
  );
}

InfiniteList.propTypes = {
  hasNextPage: PropTypes.bool.isRequired,
  isNextPageLoading: PropTypes.bool.isRequired,
  loadNextPage: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired,
  rowRender: PropTypes.func.isRequired,
  rowHeight: PropTypes.number.isRequired,
};

export default InfiniteList;
