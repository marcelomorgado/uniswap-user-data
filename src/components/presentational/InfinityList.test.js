import React from "react";
import { shallow } from "enzyme";
import InfinityList from "./InfinityList";

it("renders the component", () => {
  const props = {
    hasNextPage: false,
    isNextPageLoading: false,
    loadNextPage: () => {},
    items: [],
    rowRenderer: () => {},
    rowHeight: 10,
  };

  expect(shallow(<InfinityList {...props} />)).toMatchSnapshot();
});
