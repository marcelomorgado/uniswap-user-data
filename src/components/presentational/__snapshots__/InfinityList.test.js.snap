// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<InfiniteLoader
  isItemLoaded={[Function]}
  itemCount={0}
  loadMoreItems={[Function]}
>
  <Component />
</InfiniteLoader>
`;
